### PR TITLE
Reintroduce the Debugging types and their tests

### PR DESCRIPTION
**Pull request by tonyarnold**
Re-introduces the debugging types, including the oft-used `Debuggable` protocol.

### CODE REVIEW

**Comment by tonyarnold** [comment]
Is it worth considering pulling in [@mattgallagher's CwlDemangle](https://github.com/mattgallagher/CwlDemangle), rather than using the standard library's private demangling?

I also wasn't sure if you wanted these types and tests partitioned into their own optional module. Let me know if that's desirable.

File: Sources/Vapor/Debugging/SourceLocation.swift [hunk 1]
```diff
@@ -0,0 +1,39 @@
+/// A source-code location.
+public struct SourceLocation {
```
Comment (tanner0101) [inline Sources/Vapor/Debugging/SourceLocation.swift]: I think we have a similar type in the `AbortError` file. We should combine the two.

File: Tests/VaporTests/Debugging Tests/FooError.swift [hunk 2]
```diff
@@ -0,0 +1,55 @@
+import Vapor
```
Comment (tanner0101) [inline Tests/VaporTests/Debugging Tests/FooError.swift]: This folder should not have a space in it.

File: Tests/VaporTests/Debugging Tests/FooErrorTests.swift [hunk 3]
```diff
@@ -0,0 +1,106 @@
+import XCTest
+import Foundation
+@testable import Vapor
+
+class FooErrorTests: XCTestCase {
+    static let allTests = [
```
Comment (tanner0101) [inline Tests/VaporTests/Debugging Tests/FooErrorTests.swift]: This isn't required anymore.

File: Tests/VaporTests/Debugging Tests/FooError.swift (hunk 2 above)
Comment (tonyarnold) [reply Tests/VaporTests/Debugging Tests/FooError.swift]: Can do… but I'm casting shade at anyone who writes a script that _can't_ handle spaces in file paths. It's 2020, not 1998 😉

File: Sources/Vapor/Debugging/SourceLocation.swift (hunk 1 above)
Comment (tonyarnold) [reply Sources/Vapor/Debugging/SourceLocation.swift]: Done in 13aeb67eba0ec8ebe62f10ed7fdc7e86d8226dbd.

File: Tests/VaporTests/Debugging Tests/FooError.swift (hunk 2 above)
Comment (tonyarnold) [reply Tests/VaporTests/Debugging Tests/FooError.swift]: Done in 926a3d57f196e20b91d7d07be04effce84dd14d1.

File: Tests/VaporTests/Debugging Tests/FooErrorTests.swift (hunk 3 above)
Comment (tonyarnold) [reply Tests/VaporTests/Debugging Tests/FooErrorTests.swift]: I've removed the `alllTests` properties from the classes I added in e8ada56f6f714d0ca926f452745866bfc034177f.

**Comment by tonyarnold** [comment]
> This is great, thanks! We should integrate Debuggable into Vapor's Logger.report(error:) method so that we automatically print out this useful info.

Did you want this done as part of this PR, or as part of a follow-up? I've updated the code to downcast to `Debuggable` rather than `AbortError` for the `ErrorSource` details. 

Did you want more nuanced printing of the errors here? I'm happy to tackle it the changes if you can outline your expectations.

**Comment by tanner0101** [comment]
@tonyarnold thanks! There are a couple other things I want to tackle here including getting stacktraces working. I'm merging this into a local branch so I can work on it.

**Comment by tonyarnold** [comment]
No problems at all - glad I could take a tiny bit of the work off your plate 😄

**Pull request merged**